pkg/controller/nstemplatetier/nstemplatetier_controller_test.go

```diff
@@ -0,0 +1,575 @@
+		require.NoError(t, err2)
+		assert.NotEqual(t, hash1, hash2)
+	})
+
+}
+
+func TestReconcile(t *testing.T) {
+
+	// given
+	logf.SetLogger(logf.ZapLogger(true))
+	// a "basic" NSTemplateTier
+	oldNSTemplateTier := toolchainv1alpha1.NSTemplateTier{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: operatorNamespace,
+			Name:      "basic",
+		},
+		Spec: toolchainv1alpha1.NSTemplateTierSpec{
+			Namespaces: []toolchainv1alpha1.NSTemplateTierNamespace{
+				{
+					TemplateRef: "basic-code-123456old",
+				},
+				{
+					TemplateRef: "basic-dev-123456old",
+				},
+				{
+					TemplateRef: "basic-stage-123456old",
+				},
+			},
+			ClusterResources: &toolchainv1alpha1.NSTemplateTierClusterResources{
+				TemplateRef: "basic-clusterresources-123456old",
+			},
+		},
+	}
+	newNSTemplateTier := toolchainv1alpha1.NSTemplateTier{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: operatorNamespace,
+			Name:      "basic",
+		},
+		Spec: toolchainv1alpha1.NSTemplateTierSpec{
+			Namespaces: []toolchainv1alpha1.NSTemplateTierNamespace{
+				{
+					TemplateRef: "basic-code-123456new",
+				},
+				{
+					TemplateRef: "basic-dev-123456new",
+				},
+				{
+					TemplateRef: "basic-stage-123456new",
+				},
+			},
+			ClusterResources: &toolchainv1alpha1.NSTemplateTierClusterResources{
+				TemplateRef: "basic-clusterresources-123456new",
+			},
+		},
+	}
+	otherNSTemplateTier := toolchainv1alpha1.NSTemplateTier{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: operatorNamespace,
+			Name:      "other",
+		},
+		Spec: toolchainv1alpha1.NSTemplateTierSpec{
+			Namespaces: []toolchainv1alpha1.NSTemplateTierNamespace{
+				{
+					TemplateRef: "other-code-123456a",
+				},
+				{
+					TemplateRef: "other-dev-123456a",
+				},
+				{
+					TemplateRef: "other-stage-123456a",
+				},
+			},
+			ClusterResources: &toolchainv1alpha1.NSTemplateTierClusterResources{
+				TemplateRef: "other-clusterresources-123456a",
+			},
+		},
+	}
+
+	t.Run("controller should create a TemplateUpdateRequest", func(t *testing.T) {
+
+		// in this test, there are 10 MasterUserRecords but no associated TemplateUpdateRequest
+		t.Run("when no TemplateUpdateRequest resource exists at all", func(t *testing.T) {
+			// given
+			initObjs := []runtime.Object{&newNSTemplateTier}
+			initObjs = append(initObjs, murtest.NewMasterUserRecords(t, 10, "user-%d", murtest.Account("cluster1", oldNSTemplateTier))...)
+			r, req, cl := prepareReconcile(t, newNSTemplateTier.Name, initObjs...)
+			// when
+			res, err := r.Reconcile(req)
+			// then
+			require.NoError(t, err)
+			require.Equal(t, reconcile.Result{Requeue: true}, res) // expect a requeue to create more TemplateUpdateRequest resources
+			// check that a single TemplateUpdateRequest was created
+			actualTemplateUpdateRequests := toolchainv1alpha1.TemplateUpdateRequestList{}
+			err = cl.List(context.TODO(), &actualTemplateUpdateRequests)
+			require.NoError(t, err)
+			assert.Len(t, actualTemplateUpdateRequests.Items, 1)
```

Comment:
it's just a suggestion for simplification of all tests - you could extract this part to:
```go
func assertNumberOfTemplateUpdateRequests(t *testing.T, cl *test.FakeClient, expectedNumber int) {
	actualTemplateUpdateRequests := toolchainv1alpha1.TemplateUpdateRequestList{}
	err := cl.List(context.TODO(), &actualTemplateUpdateRequests)
	require.NoError(t, err)
	assert.Len(t, actualTemplateUpdateRequests.Items, expectedNumber)
}
```
and then use it here (and at other places too):
```go
assertNumberOfTemplateUpdateRequests(t, cl, 1)
```

Reply:
yeah, fair point

Reply:
refactored in 786f34a